File: backend/app/core/audit.py
"""
app/core/audit.py

Ghi log cho các hành động của admin (sửa giá, xóa sản phẩm, ban user, đổi trạng thái đơn...).
Dùng để truy vết khi có tranh chấp/sai sót.
"""

import logging
from datetime import datetime

logger = logging.getLogger(__name__)


def log_admin_action(
    db,
    admin: dict,
    action: str,
    target_type: str,
    target_id: str,
    details: dict | None = None,
):
    """
    Ghi 1 dòng audit log.

    admin: dict decoded từ verify_admin_token (payload JWT: sub, email, role, permissions)
    action: vd. "update_user", "delete_user", "update_order_status", "delete_product"
    target_type: vd. "user", "order", "product"
    target_id: id của đối tượng bị tác động
    details: dữ liệu bổ sung, vd. {"changes": {...}}
    """
    try:
        db.collection("audit_logs").add({
            "adminId": admin.get("sub", "unknown"),
            "adminEmail": admin.get("email", ""),
            "action": action,
            "targetType": target_type,
            "targetId": target_id,
            "details": details or {},
            "createdAt": datetime.now(),
        })
    except Exception:
        # Audit log không được phép làm fail request chính
        logger.exception(f"Không ghi được audit log: {action} {target_type}={target_id}")

File: frontend/src/lib/audit.js
/**
 * Ghi 1 bản ghi audit log vào collection "audit_logs".
 * Audit log KHÔNG được phép làm fail request chính — mọi lỗi đều bị nuốt và chỉ log ra console
 * (giống hệt nhánh `except Exception: logger.exception(...)` bên bản Python).
 *
 * @param {FirebaseFirestore.Firestore} db - dbAdmin
 * @param {object} admin - decoded JWT admin (từ requireAdmin()), cần có `sub` và `email`
 * @param {object} params
 * @param {string} params.action - vd: "create_product", "update_product", "delete_product"
 * @param {string} params.targetType - vd: "product", "order", "user"
 * @param {string} params.targetId - id của document bị tác động
 * @param {object} [params.details] - dữ liệu bổ sung (vd: diff các field đã đổi)
 */
export async function logAdminAction(db, admin, { action, targetType, targetId, details }) {
    try {
        await db.collection('audit_logs').add({
            adminId: admin?.sub || 'unknown',
            adminEmail: admin?.email || '',
            action,
            targetType,
            targetId,
            details: details || {},
            createdAt: new Date(),
        });
    } catch (err) {
        console.error(`Không ghi được audit log: ${action} ${targetType}=${targetId}`, err);
    }
}